engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -2292,73 +2292,143 @@ protected void migrate(final VMInstanceVO vm, final long srcHostId, final Deploy
         }
-        return volumeToPoolObjectMap;
     }
 
     /**
-     * We create the default mapping of volumes and storage pools for the migration of the VM to the target host.
-     * If the current storage pool of one of the volumes is using local storage in the host, it then needs to be migrated to a local storage in the target host.
-     * Otherwise, we do not need to migrate, and the volume can be kept in its current storage pool.
+     * We create the default mapping of volumes and storage pools for the migration of the VM to the target host. We execute the following checks an operations according to some scenarios:
+     * <ul>
+     *  <li>If the current storage pool of one of the volumes is using local storage in the host, it then needs to be migrated to a local storage in the target host. Otherwise, we do not need to migrate, and the volume can be kept in its current storage pool.
+     *  <li>If the current storage pool cluster is different from the target host cluster, we need to migrate the volume to one storage pool at the target cluster where the target host resides on.
+     * </ul>
      */
-    private Map<Volume, StoragePool> getDefaultMappingOfVolumesAndStoragePoolForMigration(VirtualMachineProfile profile, Host targetHost) {
+    protected Map<Volume, StoragePool> getDefaultMappingOfVolumesAndStoragePoolForMigration(VirtualMachineProfile profile, Host targetHost) {
         Map<Volume, StoragePool> volumeToPoolObjectMap = new HashMap<Volume, StoragePool>();
         List<VolumeVO> allVolumes = _volsDao.findUsableVolumesForInstance(profile.getId());
-        for (VolumeVO volume : allVolumes) {
+        createStoragePoolMappingsForVolumes(profile, targetHost, volumeToPoolObjectMap, new ArrayList<>(allVolumes));
+        return volumeToPoolObjectMap;
+    }
+
+    /**
+     * For each one of the volumes we will map it to a storage pool that is available via the target host.
+     * An exception is thrown if we cannot find a storage pool that is accessible in the target host to migrate the volume to.
+     */
+    private void createStoragePoolMappingsForVolumes(VirtualMachineProfile profile, Host targetHost, Map<Volume, StoragePool> volumeToPoolObjectMap, List<Volume> allVolumes) {
+        for (Volume volume : allVolumes) {
             StoragePoolVO currentPool = _storagePoolDao.findById(volume.getPoolId());
-            if (ScopeType.HOST.equals(currentPool.getScope())) {
-                createVolumeToStoragePoolMappingIfNeeded(profile, targetHost, volumeToPoolObjectMap, volume, currentPool);
+
+            executeManagedStorageChecks(targetHost, currentPool, volume);
+            if (ScopeType.HOST.equals(currentPool.getScope()) || isStorageCrossClusterMigration(targetHost, currentPool)) {
+                createVolumeToStoragePoolMappingIfPossible(profile, targetHost, volumeToPoolObjectMap, volume, currentPool);
             } else {
                 volumeToPoolObjectMap.put(volume, currentPool);
```

Comment:
Per a previous comment of mine, we should decide if this line should be there. The old code didn't put the volume in the map if its current pool wasn't going to change.

Reply:
Yes, explain why I am doing this in your other comment